File: src/main/java/kr/devport/api/domain/common/security/JwtTokenProvider.java
package kr.devport.api.domain.common.security;

import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.Map;

@Component
@Slf4j
public class JwtTokenProvider {

    private static final Base64.Encoder BASE64_URL_ENCODER = Base64.getUrlEncoder().withoutPadding();
    private static final Base64.Decoder BASE64_URL_DECODER = Base64.getUrlDecoder();
    private static final String HMAC_SHA512 = "HmacSHA512";
    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

    @Value("${app.jwt.secret}")
    private String jwtSecret;

    @Value("${app.jwt.access-token-expiration-ms}")
    private long accessTokenExpirationMs;

    @Value("${app.jwt.refresh-token-expiration-ms}")
    private long refreshTokenExpirationMs;

    public String generateAccessToken(Long userId) {
        return generateToken(userId, accessTokenExpirationMs);
    }

    public String createAccessToken(Long userId) {
        return generateAccessToken(userId);
    }

    public String generateRefreshToken(Long userId) {
        return generateToken(userId, refreshTokenExpirationMs);
    }

    public long getAccessTokenExpirationMs() {
        return accessTokenExpirationMs;
    }

    public long getRefreshTokenExpirationMs() {
        return refreshTokenExpirationMs;
    }

    public Long getUserIdFromToken(String token) {
        Map<String, Object> claims = parseAndValidateClaims(token);
        Object subject = claims.get("sub");
        if (!(subject instanceof String subjectValue) || subjectValue.isBlank()) {
            throw new IllegalArgumentException("JWT subject is missing");
        }
        return Long.parseLong(subjectValue);
    }

    public boolean validateToken(String token) {
        try {
            parseAndValidateClaims(token);
            return true;
        } catch (IllegalArgumentException ex) {
            log.debug("Rejected JWT token: {}", ex.getMessage());
        }
        return false;
    }

    private String generateToken(Long userId, long expirationMs) {
        long issuedAtSeconds = System.currentTimeMillis() / 1000;
        long expirationSeconds = issuedAtSeconds + (expirationMs / 1000);

        Map<String, Object> header = new LinkedHashMap<>();
        header.put("alg", "HS512");
        header.put("typ", "JWT");

        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("sub", String.valueOf(userId));
        payload.put("iat", issuedAtSeconds);
        payload.put("exp", expirationSeconds);

        String encodedHeader = encodeJson(header);
        String encodedPayload = encodeJson(payload);
        String signature = sign(encodedHeader + "." + encodedPayload);
        return encodedHeader + "." + encodedPayload + "." + signature;
    }

    private Map<String, Object> parseAndValidateClaims(String token) {
        if (token == null || token.isBlank()) {
            throw new IllegalArgumentException("JWT token is empty");
        }

        String[] parts = token.split("\\.");
        if (parts.length != 3) {
            throw new IllegalArgumentException("Invalid JWT token");
        }

        Map<String, Object> header = decodeJson(parts[0]);
        Object algorithm = header.get("alg");
        if (!"HS512".equals(algorithm)) {
            throw new IllegalArgumentException("Unsupported JWT token");
        }

        String expectedSignature = sign(parts[0] + "." + parts[1]);
        if (!MessageDigest.isEqual(
                expectedSignature.getBytes(StandardCharsets.UTF_8),
                parts[2].getBytes(StandardCharsets.UTF_8))
        ) {
            throw new IllegalArgumentException("Invalid JWT token");
        }

        Map<String, Object> claims = decodeJson(parts[1]);
        Object expiration = claims.get("exp");
        if (!(expiration instanceof Number expirationValue)) {
            throw new IllegalArgumentException("Invalid JWT token");
        }

        long nowSeconds = System.currentTimeMillis() / 1000;
        if (expirationValue.longValue() <= nowSeconds) {
            throw new IllegalArgumentException("Expired JWT token");
        }

        return claims;
    }

    private String encodeJson(Map<String, Object> value) {
        try {
            return BASE64_URL_ENCODER.encodeToString(OBJECT_MAPPER.writeValueAsBytes(value));
        } catch (Exception e) {
            throw new IllegalStateException("Failed to serialize JWT payload", e);
        }
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> decodeJson(String encodedValue) {
        try {
            byte[] decoded = BASE64_URL_DECODER.decode(encodedValue);
            return OBJECT_MAPPER.readValue(decoded, Map.class);
        } catch (Exception e) {
            throw new IllegalArgumentException("Invalid JWT token", e);
        }
    }

    private String sign(String content) {
        try {
            Mac mac = Mac.getInstance(HMAC_SHA512);
            SecretKeySpec keySpec = new SecretKeySpec(jwtSecret.getBytes(StandardCharsets.UTF_8), HMAC_SHA512);
            mac.init(keySpec);
            return BASE64_URL_ENCODER.encodeToString(mac.doFinal(content.getBytes(StandardCharsets.UTF_8)));
        } catch (Exception e) {
            throw new IllegalStateException("Failed to sign JWT token", e);
        }
    }
}

File: support/security/src/main/kotlin/kr/devport/api/domain/common/security/JwtTokenProvider.kt
package kr.devport.api.domain.common.security

import com.fasterxml.jackson.databind.ObjectMapper
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.util.Base64
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

@Component
class JwtTokenProvider(
    @param:Value("\${app.jwt.secret}") private val jwtSecret: String,
    @param:Value("\${app.jwt.access-token-expiration-ms}") val accessTokenExpirationMs: Long,
    @param:Value("\${app.jwt.refresh-token-expiration-ms}") val refreshTokenExpirationMs: Long,
) {
    private val log = LoggerFactory.getLogger(javaClass)

    fun generateAccessToken(userId: Long?): String = generateToken(userId, accessTokenExpirationMs)

    fun createAccessToken(userId: Long?): String = generateAccessToken(userId)

    fun generateRefreshToken(userId: Long?): String = generateToken(userId, refreshTokenExpirationMs)

    fun getUserIdFromToken(token: String?): Long {
        val claims = parseAndValidateClaims(token)
        val subject = claims["sub"]
        if (subject !is String || subject.isBlank()) {
            throw IllegalArgumentException("JWT subject is missing")
        }
        return subject.toLong()
    }

    fun validateToken(token: String?): Boolean {
        try {
            parseAndValidateClaims(token)
            return true
        } catch (ex: IllegalArgumentException) {
            log.debug("Rejected JWT token: {}", ex.message)
        }
        return false
    }

    private fun generateToken(
        userId: Long?,
        expirationMs: Long,
    ): String {
        val issuedAtSeconds = System.currentTimeMillis() / 1000
        val expirationSeconds = issuedAtSeconds + (expirationMs / 1000)

        val header =
            linkedMapOf<String, Any>(
                "alg" to "HS512",
                "typ" to "JWT",
            )
        val payload =
            linkedMapOf<String, Any>(
                "sub" to userId.toString(),
                "iat" to issuedAtSeconds,
                "exp" to expirationSeconds,
            )

        val encodedHeader = encodeJson(header)
        val encodedPayload = encodeJson(payload)
        val signature = sign("$encodedHeader.$encodedPayload")
        return "$encodedHeader.$encodedPayload.$signature"
    }

    private fun parseAndValidateClaims(token: String?): Map<String, Any> {
        if (token.isNullOrBlank()) {
            throw IllegalArgumentException("JWT token is empty")
        }

        val parts = token.split(".")
        if (parts.size != 3) {
            throw IllegalArgumentException("Invalid JWT token")
        }

        val header = decodeJson(parts[0])
        if (header["alg"] != "HS512") {
            throw IllegalArgumentException("Unsupported JWT token")
        }

        val expectedSignature = sign("${parts[0]}.${parts[1]}")
        if (!MessageDigest.isEqual(
                expectedSignature.toByteArray(StandardCharsets.UTF_8),
                parts[2].toByteArray(StandardCharsets.UTF_8),
            )
        ) {
            throw IllegalArgumentException("Invalid JWT token")
        }

        val claims = decodeJson(parts[1])
        val expiration = claims["exp"]
        if (expiration !is Number) {
            throw IllegalArgumentException("Invalid JWT token")
        }

        val nowSeconds = System.currentTimeMillis() / 1000
        if (expiration.toLong() <= nowSeconds) {
            throw IllegalArgumentException("Expired JWT token")
        }

        return claims
    }

    private fun encodeJson(value: Map<String, Any>): String =
        try {
            BASE64_URL_ENCODER.encodeToString(OBJECT_MAPPER.writeValueAsBytes(value))
        } catch (e: Exception) {
            throw IllegalStateException("Failed to serialize JWT payload", e)
        }

    @Suppress("UNCHECKED_CAST")
    private fun decodeJson(encodedValue: String): Map<String, Any> =
        try {
            val decoded = BASE64_URL_DECODER.decode(encodedValue)
            OBJECT_MAPPER.readValue(decoded, Map::class.java) as Map<String, Any>
        } catch (e: Exception) {
            throw IllegalArgumentException("Invalid JWT token", e)
        }

    private fun sign(content: String): String =
        try {
            val mac = Mac.getInstance(HMAC_SHA512)
            val keySpec = SecretKeySpec(jwtSecret.toByteArray(StandardCharsets.UTF_8), HMAC_SHA512)
            mac.init(keySpec)
            BASE64_URL_ENCODER.encodeToString(mac.doFinal(content.toByteArray(StandardCharsets.UTF_8)))
        } catch (e: Exception) {
            throw IllegalStateException("Failed to sign JWT token", e)
        }

    companion object {
        private val BASE64_URL_ENCODER: Base64.Encoder = Base64.getUrlEncoder().withoutPadding()
        private val BASE64_URL_DECODER: Base64.Decoder = Base64.getUrlDecoder()
        private const val HMAC_SHA512 = "HmacSHA512"
        private val OBJECT_MAPPER = ObjectMapper()
    }
}
